src/scheduler/pbs_sched.c

```diff
@@ -791,46 +502,285 @@ are_we_primary()
 	return -1;		/* cannot be neither */
 }
 
+static void
+close_server(svr_t *sconn)
+{
+	if (sconn) {
+		char *svrhost = sconn->svrhost;
+
+		delete_link(&sconn->all_svrs_link);
+		tpp_em_del_fd(poll_context, sconn->sfd);
+		pbs_disconnect(sconn->pfd);
+		pbs_disconnect(sconn->sfd);
+		if (sconn->cmd.jid)
+			free(sconn->cmd.jid);
+		free(sconn);
+		sconn = NULL;
+		log_eventf(PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_INFO,
+			   msg_daemonname, "Disconnected from server %s", svrhost ? svrhost : pbs_default());
+	}
+}
+
+static void
+close_servers(void)
+{
+	svr_t *cur, *next;
+
+	cur = GET_NEXT(servers);
+	while (cur) {
+		next = GET_NEXT(cur->all_svrs_link);
+		close_server(cur);
+		cur = next;
+	}
+	tpp_em_destroy(poll_context);
+	CLEAR_HEAD(servers);
+}
+
+static int
+connect_server_helper(char *host, int is_primary)
+{
+	int rc;
+	int fd = -1;
+	struct batch_reply *reply = NULL;
+	char *conn_type = is_primary ? "primary" : "secondary";
+
+	while (fd < 0) {
+		fd = pbs_connect(host);
+		if (fd < 0) {
+			sleep(2);
+			continue;
+		}
+	}
+
+	rc = encode_DIS_ReqHdr(fd, PBS_BATCH_RegisterSched, pbs_current_user);
+	if (rc != DIS_SUCCESS)
+		goto rerr;
+	rc = diswst(fd, sc_name);
+	if (rc != DIS_SUCCESS)
+		goto rerr;
+	rc = diswst(fd, conn_type);
+	if (rc != DIS_SUCCESS)
+		goto rerr;
+	rc = encode_DIS_ReqExtend(fd, NULL);
+	if (rc != DIS_SUCCESS)
+		goto rerr;
+	if (dis_flush(fd) != 0)
+		goto rerr;
+	pbs_errno = 0;
+	reply = PBSD_rdrpy(fd);
+	if (reply == NULL)
+		goto rerr;
+	if (pbs_errno != 0) {
+		char *errmsg = get_conn_errtxt(fd);
+		if (errmsg) {
+			log_eventf(PBSEVENT_SYSTEM | PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_NOTICE,
+				   msg_daemonname,
+				   "Server rejected register request for %s connection with error: %s", conn_type, errmsg);
+		}
+		goto rerr;
+	}
+
+	PBSD_FreeReply(reply);
+	return fd;
+
+rerr:
+	pbs_disconnect(fd);
+	PBSD_FreeReply(reply);
+	return -1;
+}
+
+static void
+connect_server(char *svrhost)
+{
+	int pfd = -1;
+	int sfd = -1;
+	svr_t *svr;
+
+	while (pfd < 0 || sfd < 0) {
+		if (pfd < 0) {
+			pfd = connect_server_helper(svrhost, 1);
+			if (pfd < 0) {
+				sleep(2);
+				continue;
+			}
+		}
+		sfd = connect_server_helper(svrhost, 0);
+		if (sfd < 0) {
+			sleep(2);
+			continue;
+		}
+	}
+	tpp_em_add_fd(poll_context, sfd, EM_IN | EM_HUP | EM_ERR);
+
+	svr = malloc(sizeof(svr_t));
+	if (!svr) {
+		log_errf(-1, __func__, "Out of memory while allocating svr struct");
```

Comment:
I understand that current code of sched is using MEM_ERR_MSG but its very generic message which is not quite helpful, saying explicit for what mem alloc failed can help more in debugging.
In fact I would suggest other way that change all MEM_ERR_MSG error in sched and put explicit message rather than very generic message.

Reply:
@bhroam does sched has separate style? if yes then why? And how does it help in search exact line of code where mem error occurs if you are using generic message?

Reply:
The scheduler has its own way of doing things that is different than the server/mom.  The daemons were written at different times for different purposes.  For example, the scheduler uses pointer arrays for ease of sorting and because the scheduler needs to have multiple arrays of the same data for different purposes (e.g. running jobs, exiting jobs, jobs in reservations that are running, etc).  Doing this with linked lists would be much much more difficult.  This is just one of the things I decided that was how the scheduler would do things early on and we always follow it.  For the scheduler, determining which malloc() fails doesn't really help.  When a malloc() fails we try and gracefully leave the cycle and hope more memory is available for the next one.  We don't care which malloc() failed, just that one did.  Being able to search for a specific string helps with that.